src/variety/tents.js

```diff
@@ -0,0 +1,744 @@
+//
+// tents.js
+//
+
+(function(pidlist, classbase) {
+	if (typeof module === "object" && module.exports) {
+		module.exports = [pidlist, classbase];
+	} else {
+		pzpr.classmgr.makeCustom(pidlist, classbase);
+	}
+})(["tents"], {
+	MouseEvent: {
+		use: true,
+		inputModes: {
+			edit: ["number", "mark-tree", "mark-tent", "shade", "clear"],
+			play: ["mark-tent", "objblank", "clear", "subline", "peke"]
+		},
+
+		mouseinput_other: function() {
+			switch (this.inputMode) {
+				case "mark-tree":
+					this.inputFixedNumber(1);
+					break;
+				case "mark-tent":
+					this.inputFixedNumber(2);
+					break;
+			}
+		},
+
+		mouseinput_number: function() {
+			if (this.mousestart) {
+				this.inputqnum_excell();
+				if (this.notInputted) {
+					this.inputqnum();
+				}
+			}
+		},
+		mouseinput_auto: function() {
+			if (this.puzzle.playmode) {
+				if (this.mousestart) {
+					this.placedTent = false;
+					this.firstPoint.reset();
+				}
+
+				var cell = this.getcell();
+
+				if (cell.isnull) {
+					return;
+				}
+				if (this.inputData === null && !this.firstPoint.oncell()) {
+					this.firstPoint.set(cell);
+				}
+
+				if (this.inputData === null && !this.firstPoint.equals(cell)) {
+					var other = this.firstPoint.getc();
+					var border = cell.getnb(other);
+					var single = this.puzzle.getConfig("use") === 1;
+
+					var hastree = other.getNum() === 1 || cell.getNum() === 1;
+					var hastent = other.getNum() === 2 || cell.getNum() === 2;
+					var hasdot =
+						other.getNum() === 3 ||
+						other.qsub === 1 ||
+						cell.getNum() === 3 ||
+						cell.qsub === 1;
+					var hasempty =
+						(other.getNum() === -1 && other.qsub === 0) ||
+						(cell.getNum() === -1 && cell.qsub === 0);
+
+					if (hastree && hastent) {
+						this.inputData = 1;
+					}
+
+					if (hastree && hasempty) {
+						if (!border.isnull && border.qsub === 1) {
+							this.inputData = 0;
+						} else {
+							this.inputData = this.btn === "left" ? 1 : 2;
+						}
+					}
+
+					if (hastree && hasdot) {
+						this.inputData = this.btn === "left" && !single ? 1 : 2;
+					}
+
+					this.inputcell_tents(other);
```

Comment:
I'm trying to understand why the starting cell is only input with delay. I guess this is the point where that happens?

E.g.:

1. left mouse down on cell (nothing happens)
2. drag to neighbour cell (tree appears in starting cell)
3. lift mouse button (nothing happens)

Similar behaviour occurs for dots. In both cases I'd expect the symbol to appear on first mouse-down, but I may be missing some reason why we can't know for sure that's going to happen at that point?

Reply:
The delay is to differentiate between cell input and border input. Without a delay, you can't drag a line from an existing tent to a tree without erasing the tent. This is more noticable when you only want to remove a line without removing the tent. There's a lot of special edge-cases which are covered by the unit tests.

To get something to happen on mousedown, I recommend not actually changing the input order, but maybe set an additional value on the first cell (such as `info`) which allows for purely visual feedback. Mouseup would clear this value on all cells. You could use this to change the cell background to a special color like in Shugaku (School Trip), or to draw a tent when the actual cell is still empty.

Reply:
Right-clicking an empty cell now inputs a dot on mousedown, except when the mouse is on a border. Left-clicking an empty cell will display a fake tent on mousedown using `setinfo` as described above.

All actions still work the same as before, but now there's feedback on mousedown in most common cases.